interface/ceed-basis.c

```diff
@@ -883,6 +836,86 @@ int CeedBasisCreateHdiv(Ceed ceed, CeedElemTopology topo, CeedInt num_comp, Ceed
 }
 
+/**
+  @brief Create a non tensor-product basis for H(curl) discretizations
+
+  @param[in]  ceed      Ceed object where the CeedBasis will be created
+  @param[in]  topo      Topology of element (`CEED_TOPOLOGY_QUAD`, `CEED_TOPOLOGY_PRISM`, etc.), dimension of which is used in some array sizes below
+  @param[in]  num_comp  Number of components (usually 1 for vectors in H(curl) bases)
+  @param[in]  num_nodes Total number of nodes (dofs per element)
+  @param[in]  num_qpts  Total number of quadrature points
+  @param[in]  interp    Row-major (dim * num_qpts * num_nodes) matrix expressing the values of basis functions at quadrature points
+  @param[in]  curl      Row-major (cdim * num_qpts * num_nodes, cdim = 1 if dim < 3 else dim) matrix expressing curl of basis functions at quadrature
+points
+  @param[in]  q_ref     Array of length num_qpts * dim holding the locations of quadrature points on the reference element
+  @param[in]  q_weight  Array of length num_qpts holding the quadrature weights on the reference element
+  @param[out] basis     Address of the variable where the newly created CeedBasis will be stored.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedBasisCreateHcurl(Ceed ceed, CeedElemTopology topo, CeedInt num_comp, CeedInt num_nodes, CeedInt num_qpts, const CeedScalar *interp,
+                         const CeedScalar *curl, const CeedScalar *q_ref, const CeedScalar *q_weight, CeedBasis *basis) {
+  CeedInt Q = num_qpts, P = num_nodes, dim = 0, cdim;
+  CeedCall(CeedBasisGetTopologyDimension(topo, &dim));
+  cdim = (dim < 3) ? 1 : dim;
+  if (!ceed->BasisCreateHdiv) {
+    Ceed delegate;
+    CeedCall(CeedGetObjectDelegate(ceed, &delegate, "Basis"));
+
+    if (!delegate) {
+      // LCOV_EXCL_START
+      return CeedError(ceed, CEED_ERROR_UNSUPPORTED, "Backend does not implement BasisCreateHcurl");
+      // LCOV_EXCL_STOP
+    }
+
+    CeedCall(CeedBasisCreateHcurl(delegate, topo, num_comp, num_nodes, num_qpts, interp, curl, q_ref, q_weight, basis));
+    return CEED_ERROR_SUCCESS;
+  }
+
+  if (num_comp < 1) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_DIMENSION, "Basis must have at least 1 component");
+    // LCOV_EXCL_STOP
+  }
+
+  if (num_nodes < 1) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_DIMENSION, "Basis must have at least 1 node");
+    // LCOV_EXCL_STOP
+  }
+
+  if (num_qpts < 1) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_DIMENSION, "Basis must have at least 1 quadrature point");
+    // LCOV_EXCL_STOP
+  }
+
+  CeedCall(CeedCalloc(1, basis));
+
+  (*basis)->ceed = ceed;
+  CeedCall(CeedReference(ceed));
+  (*basis)->ref_count    = 1;
+  (*basis)->tensor_basis = 0;
+  (*basis)->dim          = dim;
+  (*basis)->topo         = topo;
+  (*basis)->num_comp     = num_comp;
+  (*basis)->P            = P;
+  (*basis)->Q            = Q;
+  (*basis)->fe_space     = 3;  // H(curl) space
```

Comment:
Should use the Enum value here instead of hard-coding the magic number.

Reply:
Resolved in [063bea1](https://github.com/CEED/libCEED/pull/1156/commits/063bea108b882e83a5ae894d5a49e974101cbd30).